File: v2/static/js/evaluate_finetune.js
/* Evaluate Fine-tune — comparison logic */

let _compareAbort = null;

async function runComparison() {
  const btn = document.getElementById("compareBtn");
  const progress = document.getElementById("compareProgress");
  const bar = document.getElementById("cmpBar");
  const lbl = document.getElementById("cmpLabel");
  const results = document.getElementById("compareResults");
  const orig = btn.innerHTML;

  btn.disabled = true;
  btn.innerHTML = '<span class="spin-icon"></span>Running both models…';
  progress.style.display = "block";
  results.style.display = "none";
  _compareAbort = new AbortController();

  const steps = [
    { pct: 10, label: "Sending prompt to both models…" },
    { pct: 35, label: "Production model generating…" },
    { pct: 55, label: "Fine-tuned model generating…" },
    { pct: 80, label: "Scoring & validating candidates…" },
    { pct: 95, label: "Building comparison report…" },
  ];
  let si = 0;
  const timer = setInterval(() => {
    if (si < steps.length) {
      bar.style.width = steps[si].pct + "%";
      lbl.textContent = steps[si].label;
      si++;
    }
  }, 2800);

  try {
    const payload = {
      smile: document.getElementById("seed_smiles").value.trim(),
      amino_acid_seq: document.getElementById("amino_acid_seq").value.trim(),
      num_candidates: parseInt(document.getElementById("num_candidates").value),
      noise: parseFloat(document.getElementById("noise_level").value),
      target_id: document.getElementById("target_id").value,
      target_name: document.getElementById("target_name_val").value,
      uniprot_id: document.getElementById("uniprot_id").value,
      pdb_id: document.getElementById("pdb_id").value,
    };

    const data = await apiFetch("/api/v2/compare-models", {
      method: "POST",
      body: JSON.stringify(payload),
      signal: _compareAbort.signal,
    });

    clearInterval(timer);
    bar.style.width = "100%";
    lbl.textContent = "Done!";

    renderResults(data);
    results.style.display = "block";
    results.scrollIntoView({ behavior: "smooth", block: "start" });

  } catch (e) {
    clearInterval(timer);
    if (e.name !== "AbortError") {
      bar.style.width = "0%";
      lbl.textContent = "Error: " + (e.message || "Unknown error");
      showAlert("Comparison failed: " + (e.message || "check server logs"), "danger");
    }
  } finally {
    btn.disabled = false;
    btn.innerHTML = orig;
    _compareAbort = null;
  }
}

function renderResults(data) {
  renderModelPanel("prod", data.production);
  renderModelPanel("ft",   data.finetuned);
  renderReport(data.comparison, data.production, data.finetuned);
}

function renderModelPanel(key, model) {
  const s = model.stats;
  const tableEl = document.getElementById(`${key}Table`);
  const statsEl = document.getElementById(`${key}Stats`);
  const errEl   = document.getElementById(`${key}Error`);

  if (errEl) errEl.style.display = model.error && !model.candidates.length ? "block" : "none";
  if (errEl && model.error) errEl.textContent = "Error: " + model.error;

  statsEl.innerHTML = `
    <div class="stat-row"><span class="label">Candidates found</span><span class="val">${s.valid_count}</span></div>
    <div class="stat-row"><span class="label">Avg QED</span><span class="val">${s.avg_qed.toFixed(3)}</span></div>
    <div class="stat-row"><span class="label">Avg SAS</span><span class="val">${s.avg_sas.toFixed(2)}</span></div>
    <div class="stat-row"><span class="label">Avg Score</span><span class="val">${(s.avg_composite * 100).toFixed(0)}%</span></div>
    <div class="stat-row"><span class="label">Latency</span><span class="val">${(model.latency_ms / 1000).toFixed(1)}s</span></div>
    <div class="stat-row"><span class="label">Uniqueness</span><span class="val">${(s.unique_rate * 100).toFixed(0)}%</span></div>
  `;

  if (!model.candidates.length) {
    tableEl.innerHTML = "<p style='color:var(--text-muted);padding:8px 0;font-size:.83rem;'>No valid candidates.</p>";
    return;
  }

  const SCORE_CLR = s => s >= 0.7 ? "#3fb950" : s >= 0.4 ? "#ffa657" : "#ff7b72";
  const rows = model.candidates.map(c => {
    const short = c.smiles.length > 22 ? c.smiles.slice(0, 22) + "…" : c.smiles;
    return `<tr onclick="copySmiles('${c.smiles.replace(/'/g, "\\'")}')" title="Click to copy SMILES">
      <td>${c.rank}</td>
      <td style="font-family:monospace;font-size:.72rem;">${short}</td>
      <td style="color:${SCORE_CLR(c.composite_score)};font-weight:700;">${(c.composite_score * 100).toFixed(0)}%</td>
      <td>${c.qed != null ? c.qed.toFixed(2) : "—"}</td>
      <td>${c.sas != null ? c.sas.toFixed(1) : "—"}</td>
      <td>${c.lipinski_pass ? "✅" : "❌"}</td>
    </tr>`;
  }).join("");

  tableEl.innerHTML = `
    <div style="overflow-x:auto;">
      <table class="cmp-candidate-table">
        <thead><tr><th>#</th><th>SMILES</th><th>Score</th><th>QED</th><th>SAS</th><th>Lipinski</th></tr></thead>
        <tbody>${rows}</tbody>
      </table>
    </div>`;
}

function renderReport(cmp, prod, ft) {
  const banner = document.getElementById("winnerBanner");
  const metricsEl = document.getElementById("metricGrid");

  const names = { production: "Production (dlyog04)", finetuned: "Fine-tuned (dgx1)", tie: "Tie" };
  const cls   = { production: "winner-production",    finetuned: "winner-finetuned",   tie: "winner-tie" };

  banner.className = "report-winner-banner " + cls[cmp.overall_winner];
  banner.innerHTML = `Overall Winner: <strong>${names[cmp.overall_winner]}</strong>
    <span style="font-size:.8rem;font-weight:400;margin-left:12px;">
      Prod ${(cmp.production_score * 100).toFixed(0)}% &nbsp;|&nbsp; FT ${(cmp.finetuned_score * 100).toFixed(0)}%
    </span>`;

  const metrics = [
    { name: "SMILES Validity",  key: "validity",    prodVal: cmp.validity_prod + "%",     ftVal: cmp.validity_ft + "%" },
    { name: "Avg QED",          key: "avg_qed",     prodVal: prod.stats.avg_qed.toFixed(3), ftVal: ft.stats.avg_qed.toFixed(3) },
    { name: "Avg SAS",          key: "avg_sas",     prodVal: prod.stats.avg_sas.toFixed(2), ftVal: ft.stats.avg_sas.toFixed(2) },
    { name: "Avg Score",        key: "composite",   prodVal: (prod.stats.avg_composite * 100).toFixed(0) + "%", ftVal: (ft.stats.avg_composite * 100).toFixed(0) + "%" },
    { name: "Uniqueness",       key: "unique_rate", prodVal: (prod.stats.unique_rate * 100).toFixed(0) + "%",   ftVal: (ft.stats.unique_rate * 100).toFixed(0) + "%" },
    { name: "Speed",            key: "latency",     prodVal: (prod.latency_ms / 1000).toFixed(1) + "s",         ftVal: (ft.latency_ms / 1000).toFixed(1) + "s" },
  ];

  const winCls = { production: "prod", finetuned: "ft", tie: "tie" };
  const winLabel = { production: "Prod wins", finetuned: "FT wins", tie: "Tie" };

  metricsEl.innerHTML = metrics.map(m => {
    const w = cmp.wins[m.key];
    return `<div class="metric-card">
      <div class="metric-name">${m.name}</div>
      <div class="metric-winner ${winCls[w]}">${winLabel[w]}</div>
      <div class="metric-vals">Prod: ${m.prodVal} &nbsp; FT: ${m.ftVal}</div>
    </div>`;
  }).join("");

  document.getElementById("reportCard").style.display = "block";
}

function copySmiles(smiles) {
  navigator.clipboard.writeText(smiles).then(() => {
    showAlert("SMILES copied to clipboard!", "success");
  }).catch(() => {
    showAlert("Copy failed — SMILES: " + smiles, "info");
  });
}

async function loadModelHealth() {
  try {
    const data = await apiFetch("/api/v2/model-health");
    _setHealthDot("prodDot", data.production);
    _setHealthDot("ftDot",   data.finetuned);
    _setHealthLabel("prodStatus", data.production, "dlyog04:9000");
    _setHealthLabel("ftStatus",   data.finetuned,  "dgx1:9002");
  } catch (_) {}
}

function _setHealthDot(id, up) {
  const el = document.getElementById(id);
  if (!el) return;
  el.className = "health-dot " + (up ? "up" : "down");
  el.title = up ? "Online" : "Offline";
}

function _setHealthLabel(id, up, host) {
  const el = document.getElementById(id);
  if (!el) return;
  el.textContent = host + " — " + (up ? "Online" : "Offline");
  el.style.color = up ? "#3fb950" : "#ff7b72";
}

File: v2/api/evaluate_finetune.py
"""Evaluate Fine-tune — side-by-side comparison of production vs fine-tuned model."""
import statistics
from flask import Blueprint, jsonify, render_template, request
from flask_login import login_required, current_user
from marshmallow import ValidationError

from api.schemas import CompareSchema
from services import molecule_generator, property_calculator, dti_predictor
from services.target_service import get_curated_target
from services.molecule_validator import validate_and_canonicalize
from utils.protein_utils import is_valid_sequence, clean_sequence

bp = Blueprint("evaluate_finetune", __name__)


@bp.route("/evaluate-finetune")
@login_required
def evaluate_page():
    target_id = request.args.get("target_id", "").strip()
    target = get_curated_target(target_id) if target_id else None
    return render_template("evaluate_finetune.html", target=target)


@bp.route("/api/v2/compare-models", methods=["POST"])
@login_required
def compare_models():
    try:
        data = CompareSchema().load(request.get_json() or {})
    except ValidationError as e:
        return jsonify({"error": e.messages}), 400

    aa_seq = clean_sequence(data["amino_acid_seq"])
    if aa_seq and not is_valid_sequence(aa_seq):
        return jsonify({"error": "Invalid amino acid sequence"}), 400

    canon_seed = validate_and_canonicalize(data["smile"])
    if canon_seed is None:
        return jsonify({"error": "Invalid seed SMILES"}), 400

    try:
        both = molecule_generator.generate_both(
            seed_smile=canon_seed,
            amino_acid_seq=aa_seq,
            noise=data["noise"],
            n=data["num_candidates"],
        )
    except RuntimeError as e:
        return jsonify({"error": str(e)}), 503

    return jsonify({
        "production": _score_results(both["production"], canon_seed, aa_seq),
        "finetuned":  _score_results(both["finetuned"],  canon_seed, aa_seq),
        "comparison": _compare(
            _score_results(both["production"], canon_seed, aa_seq),
            _score_results(both["finetuned"],  canon_seed, aa_seq),
        ),
    }), 200


@bp.route("/api/v2/model-health")
@login_required
def model_health():
    return jsonify({
        "production":   molecule_generator.check_dgx_health(),
        "finetuned":    molecule_generator.check_finetuned_health(),
        "finetuned_v2": molecule_generator.check_finetuned_v2_health(),
    })


def _score_results(gen: dict, canon_seed: str, aa_seq: str) -> dict:
    if gen.get("error") and not gen["smiles"]:
        return {
            "candidates": [], "latency_ms": 0, "error": gen["error"],
            "stats": {"valid_count": 0, "avg_qed": 0, "avg_sas": 0,
                      "avg_composite": 0, "unique_rate": 0},
        }

    candidates = []
    for rank, smiles in enumerate(gen["smiles"], start=1):
        props = property_calculator.compute_all(smiles, canon_seed)
        if props is None:
            continue
        dti = dti_predictor.predict(props, aa_seq)
        comp = dti_predictor.composite_score(props, dti)
        candidates.append({
            "rank": rank,
            "smiles": smiles,
            "composite_score": round(comp, 3),
            "dti_score": round(dti, 3),
            **{k: (round(v, 3) if isinstance(v, float) else v) for k, v in props.items()},
        })

    candidates.sort(key=lambda c: c["composite_score"], reverse=True)
    for i, c in enumerate(candidates, 1):
        c["rank"] = i

    n = len(candidates)
    stats = {
        "valid_count": n,
        "avg_qed": round(statistics.mean(c["qed"] for c in candidates), 3) if n else 0,
        "avg_sas": round(statistics.mean(c["sas"] for c in candidates), 3) if n else 0,
        "avg_composite": round(statistics.mean(c["composite_score"] for c in candidates), 3) if n else 0,
        "unique_rate": round(n / max(gen["total_generated"], 1), 3),
    }

    return {
        "candidates": candidates,
        "latency_ms": gen["latency_ms"],
        "total_parsed": gen["total_generated"],
        "stats": stats,
        "error": gen.get("error"),
    }


def _compare(prod: dict, ft: dict) -> dict:
    ps = prod["stats"]
    fs = ft["stats"]

    def winner(prod_val, ft_val, higher_better=True):
        if prod_val == ft_val:
            return "tie"
        if higher_better:
            return "production" if prod_val > ft_val else "finetuned"
        return "production" if prod_val < ft_val else "finetuned"

    validity_prod = ps["valid_count"] / max(prod.get("total_parsed", 1), 1)
    validity_ft   = fs["valid_count"] / max(ft.get("total_parsed", 1), 1)

    wins = {
        "validity":    winner(validity_prod, validity_ft),
        "avg_qed":     winner(ps["avg_qed"],     fs["avg_qed"]),
        "avg_sas":     winner(ps["avg_sas"],      fs["avg_sas"],      higher_better=False),
        "latency":     winner(prod["latency_ms"], ft["latency_ms"],   higher_better=False),
        "unique_rate": winner(ps["unique_rate"],  fs["unique_rate"]),
        "composite":   winner(ps["avg_composite"], fs["avg_composite"]),
    }

    weights = {"validity": 0.30, "avg_qed": 0.25, "avg_sas": 0.15,
               "composite": 0.15, "unique_rate": 0.10, "latency": 0.05}

    prod_score = sum(w for k, w in weights.items() if wins[k] == "production")
    ft_score   = sum(w for k, w in weights.items() if wins[k] == "finetuned")

    overall = "production" if prod_score > ft_score else ("finetuned" if ft_score > prod_score else "tie")

    return {
        "wins": wins,
        "overall_winner": overall,
        "production_score": round(prod_score, 2),
        "finetuned_score":  round(ft_score, 2),
        "validity_prod": round(validity_prod * 100, 1),
        "validity_ft":   round(validity_ft * 100, 1),
    }
